redis/src/tls.rs

```diff
@@ -0,0 +1,174 @@
+/// - CA's certificate
+/// - client's certificate and
+/// - client's key
+///
+/// Example usage::
+///
+/// ```rust,no_run
+/// let mut tls_builder = TLSClientBuilder::new(server.to_string(), port);
+/// let cert_file = File::open(cert_file).expect("cannot open private cert file");
+/// let mut reader = BufReader::new(cert_file);
+/// tls_builder
+///     .client_certs(&mut reader)
+///     .expect("Unable to handle client certs");
+/// let key_file = File::open(key_file).expect("Cannot open private key file");
+/// let mut reader = BufReader::new(key_file);
+/// tls_builder
+///     .client_key(&mut reader)
+///     .expect("Unable to handle client key");
+/// let ca_file = File::open(ca_file).expect("Cannot open CA cert file");
+/// let mut reader = BufReader::new(ca_file);
+/// tls_builder
+///     .root_cert(&mut reader)
+///     .expect("Unable to load CA cert");
+///
+/// let client = tls_builder.build().expect("Unable to build client");
+/// ```
+///
+impl TLSClientBuilder {
+    /// Initiate builder with given _server_ and _port_
+    pub fn new(host: String, port: u16) -> TLSClientBuilder {
+        TLSClientBuilder {
+            host,
+            port,
+            client_key: None,
+            client_certs: Vec::new(),
+            root_cert: RootCertStore::empty(),
+        }
+    }
+
+    /// Build a TLS client with parameters provided to builder:
+    ///
+    /// If all CA certificate, Client Certificate and Client key are provided,
+    /// it builds a TLS client with mutual authentication
+    ///
+    /// If none of certificates and key are *not* provided, it builds a TLS
+    /// client with *root* CA certificate
+    ///
+    /// Otherwise it returns an error
+    pub fn build(self) -> RedisResult<Client> {
+        let tls_params = {
+            match (self.client_key, self.client_certs, self.root_cert) {
+                (Some(client_key), client_certs, root_cert) => Some(TlsConnParams {
+                    client_key,
+                    client_certs,
+                    root_cert,
+                }),
+                (None, client_certs, root_cert)
+                    if client_certs.is_empty() && root_cert.is_empty() =>
+                {
+                    None
+                }
+                _ => fail!((
+                    ErrorKind::InvalidClientConfig,
+                    "TLS client config not ready"
+                )),
+            }
+        };
+
+        Ok(Client {
+            connection_info: ConnectionInfo {
+                addr: ConnectionAddr::TcpTls {
+                    host: self.host,
+                    port: self.port,
+                    insecure: false,
+                    tls_params,
+                },
+                redis: RedisConnectionInfo::default(),
+            },
+        })
+    }
+
+    /// Loads a client certificate in PEM format as a byte stream
+    /// In case of errors loading the certificate it returns an error
```

Comment:
doc nit:

```suggestion
    /// Loads a client certificate in PEM format as a byte stream.
    ///
    /// In case of errors loading the certificate it returns an error.
```